inotify_poller.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2015 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// +build linux
+
+package fsnotify
+
+import (
+	"errors"
+	"os"
+	"syscall"
+)
+
+type fdPoller struct {
+	fd   int    // File descriptor (as returned by the inotify_init() syscall)
+	epfd int    // Epoll file descriptor
+	pipe [2]int // Pipe for waking up
+}
+
+// Create a new inotify poller.
+// This creates an inotify handler, and an epoll handler.
+func newFdPoller() (*fdPoller, error) {
+	var errno error
+	poller := new(fdPoller)
+
+	// Create inotify fd
+	poller.fd, errno = syscall.InotifyInit()
+	if poller.fd == -1 {
+		return nil, os.NewSyscallError("inotify_init", errno)
+	}
+	// Create epoll fd
+	poller.epfd, errno = syscall.EpollCreate(1)
+	if poller.epfd == -1 {
+		syscall.Close(poller.fd)
+		return nil, os.NewSyscallError("epoll_create", errno)
+	}
+	// Create pipe; pipe[0] is the read end, pipe[1] the write end.
+	errno = syscall.Pipe(poller.pipe[:])
+	if errno != nil {
+		syscall.Close(poller.fd)
+		syscall.Close(poller.epfd)
+		return nil, os.NewSyscallError("pipe", errno)
+	}
+
+	// Register inotify fd with epoll
+	event := syscall.EpollEvent{
+		Fd:     int32(poller.fd),
+		Events: syscall.EPOLLIN,
+	}
+	errno = syscall.EpollCtl(poller.epfd, syscall.EPOLL_CTL_ADD, poller.fd, &event)
+	if errno != nil {
+		syscall.Close(poller.fd)
+		syscall.Close(poller.epfd)
+		syscall.Close(poller.pipe[0])
+		syscall.Close(poller.pipe[1])
+		return nil, os.NewSyscallError("epoll_ctl", errno)
+	}
+
+	// Register pipe fd with epoll
+	event = syscall.EpollEvent{
+		Fd:     int32(poller.pipe[0]),
+		Events: syscall.EPOLLIN,
+	}
+	errno = syscall.EpollCtl(poller.epfd, syscall.EPOLL_CTL_ADD, poller.pipe[0], &event)
+	if errno != nil {
+		syscall.Close(poller.fd)
```

Comment:
maybe cleanup could be put in a `defer func() {}` that checks if err is != nil? Might be cleaner, maybe not.

Maybe extracting a few functions, like to create and close the pipe, would help a bit?